crates/bevy_ecs/src/event/peek_iterators.rs

```diff
@@ -0,0 +1,240 @@
+use crate as bevy_ecs;
+use bevy_ecs::{
+    batching::BatchingStrategy,
+    event::{Event, EventId, EventInstance, Events, ManualEventReader},
+};
+use bevy_utils::detailed_trace;
+use std::{iter::Chain, slice::Iter};
+
+/// An iterator that yields any unread events from an [`EventReader`] or [`ManualEventReader`].
+#[derive(Debug)]
+pub struct EventPeekIterator<'a, E: Event> {
+    iter: EventPeekIteratorWithId<'a, E>,
+}
+
+impl<'a, E: Event> Iterator for EventPeekIterator<'a, E> {
+    type Item = &'a E;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next().map(|(event, _)| event)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
+    }
+
+    fn count(self) -> usize {
+        self.iter.count()
+    }
+
+    fn last(self) -> Option<Self::Item>
+    where
+        Self: Sized,
+    {
+        self.iter.last().map(|(event, _)| event)
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        self.iter.nth(n).map(|(event, _)| event)
+    }
+}
+
+impl<'a, E: Event> ExactSizeIterator for EventPeekIterator<'a, E> {
+    fn len(&self) -> usize {
+        self.iter.len()
+    }
+}
+
+/// An iterator that yields references to any unread events (and their IDs) from an [`EventReader`] or [`ManualEventReader`].
+#[derive(Debug)]
+pub struct EventPeekIteratorWithId<'a, E: Event> {
+    chain: Chain<Iter<'a, EventInstance<E>>, Iter<'a, EventInstance<E>>>,
+    unread: usize,
+}
+
+impl<'a, E: Event> EventPeekIteratorWithId<'a, E> {
+    /// Creates a new iterator that yields any `events` that have not yet been seen by `reader`.
+    pub fn new(reader: &'a ManualEventReader<E>, events: &'a Events<E>) -> Self {
+        let a_index = reader
+            .last_event_count
+            .saturating_sub(events.events_a.start_event_count);
+        let b_index = reader
+            .last_event_count
+            .saturating_sub(events.events_b.start_event_count);
+        let a = events.events_a.get(a_index..).unwrap_or_default();
+        let b = events.events_b.get(b_index..).unwrap_or_default();
+
+        let unread = a.len() + b.len();
+        // Ensure `len` is implemented correctly
+        debug_assert_eq!(unread, reader.len(events));
+
+        // Iterate the oldest first, then the newer events
+        let chain = a.iter().chain(b.iter());
+
+        Self { chain, unread }
+    }
+
+    /// Iterate over only the events.
```

Comment:
```suggestion
    /// Iterate over the events only.
```
(Unsure if this is correct though, it just sounds better to me)